include/picongpu/plugins/binning/DomainInfo.hpp

```diff
@@ -0,0 +1,54 @@
+/* Copyright 2023 Tapish Narwal
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
```

Comment:
I'm having some trouble compiling this plugin at the moment since some `.hpp` files (this one, but also others like `UnitConversion.hpp`) have no `#include` directives, so only a very specific include order will work.
Can you give a short guide on how to compile and run a working instance of this plugin?

Reply:
could you please post the compiler error log that we can trace from where the error starts.

Reply:
It's not only one file that is affected by this:

```
[ 97%] Building CUDA object CMakeFiles/picongpu.dir/versionFormat.cpp.o                                                                                                                                                                                                                                                    
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(72): error: the global scope has no "openPMD"                                                                                                                                                                       
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(72): error: too few arguments for class template "std::map"                                                                                                                                                         
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(72): error: expected an identifier                                                                                                                                                                                  
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(74): error: the global scope has no "openPMD"                                                                                                                                                                       
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(74): error: expected a ";"                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name                                                                                                                                                 
                                                                                                                                                                                                                                                                                                                           
/home/franzpoeschel/git-repos/picongpu/include/picongpu/../picongpu/plugins/binning/UnitConversion.hpp(76): error: name followed by "::" must be a class or namespace name  
```

When I fix the includes here, the next include error is in `DomainInfo.hpp`.

Error triggered by adding `extensionParam.loader` with below content to LaserWakefield example and trying to compile it:

```cpp
#pragma once
#include "picongpu/param/binningSetup.param"
```

Reply:
This plugin is different compared to our old plugins. We kept the configuration in the param folder but this file is not allowed to be included from `param.loader` or the extension.

Reply:
Also, could you pull again before trying again :smile:

Reply:
I pulled again. How do I run this then? There is nothing in `picongpu --help` except for the single line `Binning Dispatcher: A flexible binning plugin!:`

Reply:
Or are this the `--e_energyHistogram.*` parameters?

Reply:
So there is no runtime configuration, and to run the plugin, the user config goes into `picongpu/param/binningSetup.param` . There is an existing example in that file which you can simply call in the `getBinning()` function as `binningExample(binningCreator)`

Reply:
Ah. I also had a simulation where the `if constexpr` part in the example apparently evaluated to `false`, so nothing happened. I have it running now.

Reply:
I realized that it was always evaluating to false because i was using `mpl::true_` instead of `mp11::mp_true` :sweat_smile: